src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.cpp

```diff
@@ -0,0 +1,1359 @@
+  auto &body = goto_function.body;
+
+  // rewrite is_fresh_calls
+  dfcc_is_fresht is_fresh(library, log);
+  is_fresh.rewrite_calls(body, null_expr);
+
+  // rewrite is_freeable/was_freed calls
+  dfcc_is_freeablet is_freeable(library, log);
+  is_freeable.rewrite_calls(body, null_expr);
+
+  // rewrite calls
+  Forall_goto_program_instructions(it, body)
+  {
+    if(it->is_function_call())
+      instrument_call_instruction(null_expr, it, body);
+  }
+
+  goto_model.goto_functions.update();
+
+  // debug print
+  forall_goto_program_instructions(it, body)
+  {
+    body.output_instruction(ns, function_id, log.debug(), *it);
+  }
+  log.debug() << "<-dfcc_instrumentt::instrument_harness_function("
+              << function_id << ")" << messaget::eom;
+}
+
+void dfcc_instrumentt::collect_local_statics(
+  const irep_idt &function_id,
+  std::set<symbol_exprt> &local_statics)
+{
+  for(const auto &sym_pair : goto_model.symbol_table)
+  {
+    const auto &sym = sym_pair.second;
+    if(sym.is_static_lifetime)
+    {
+      const auto &loc = sym.location;
+      if(!loc.get_function().empty() && loc.get_function() == function_id)
+      {
+        local_statics.insert(sym.symbol_expr());
+      }
+    }
+  }
+}
+
+void dfcc_instrumentt::instrument_function(const irep_idt &function_id)
+{
+  // use same name for local static search
+  instrument_function(function_id, function_id);
+}
+
+void dfcc_instrumentt::instrument_wrapped_function(
+  const irep_idt &wrapped_function_id,
+  const irep_idt &initial_function_id)
+{
+  // use the initial name name for local static search
+  instrument_function(wrapped_function_id, initial_function_id);
+}
+
+void dfcc_instrumentt::instrument_function(
+  const irep_idt &function_id,
+  const irep_idt &function_id_for_local_static_search)
+{
+  log.debug() << "->dfcc_instrumentt::instrument_function(" << function_id
+              << ")" << messaget::eom;
+
+  // never instrument a function twice
+  if(
+    dfcc_instrumentt::function_cache.find(function_id) !=
+    dfcc_instrumentt::function_cache.end())
+    return;
+
+  dfcc_instrumentt::function_cache.insert(function_id);
+
+  auto found = goto_model.goto_functions.function_map.find(function_id);
+  PRECONDITION_WITH_DIAGNOSTICS(
+    found != goto_model.goto_functions.function_map.end(),
+    "Function '" + id2string(function_id) + "' must exist in the model.");
+  // TODO instead of erroring out, create an entry in the function table
+  // and create body with assert(false); assume(false)
+
+  auto &goto_function = found->second;
+
+  function_cfg_infot cfg_info(goto_function);
+
+  const auto &write_set = utils.add_parameter(
+    function_id,
+    "__write_set_to_check",
+    library.dfcc_type[dfcc_typet::WRITE_SET_PTR]);
+
+  std::set<symbol_exprt> local_statics;
+  collect_local_statics(function_id_for_local_static_search, local_statics);
+
+  instrument_function_body(
+    function_id, write_set.symbol_expr(), cfg_info, local_statics);
```

Comment:
since from the function id you can always get the goto function, goto program, or the symbol, I'm using function_ids everywhere to make interfaces uniform. Internally each function fetches what it needs to work on.